aten/src/ATen/native/Distributions.h

```diff
@@ -0,0 +1,215 @@
+      return scale * d * v;
+  }
+}
+
+template <typename scalar_t>
+deviceforcuda static inline scalar_t polevl(const scalar_t x,  const scalar_t A[], size_t len) {
+  scalar_t result = 0;
+  for (size_t i = 0; i <= len; i++) {
+    result = result * x + A[i];
+  }
+  return result;
+}
+
+
+/*
+ * The following function comes with the following copyright notice.
+ * It has been released under the BSD license.
+ *
+ * Cephes Math Library Release 2.8:  June, 2000
+ * Copyright 1984, 1987, 1992, 2000 by Stephen L. Moshier
+ */
+template <typename scalar_t>
+deviceforcuda static inline scalar_t digamma_one(scalar_t x) {
+  using acc_scalar_t = typename std::conditional<std::is_same<scalar_t, at::Half>::value, float, scalar_t>::type;
+  constexpr acc_scalar_t PSI_10 = 2.25175258906672110764;
+  if (x == 0) {
+    return INFINITY;
+  }
+  acc_scalar_t additional_summand = 0;
+  int x_is_integer = x == std::floor(x);
+  if (x < 0) {
+    if (x_is_integer) {
+      return INFINITY;
+    }
+    // it is more standard to write this as recursion, but
+    // nvcc does not like that
+    additional_summand = - M_PI / std::tan(M_PI * x);
+    x = 1 - x;
+  }
+
+  // Push x to be >= 10
+  acc_scalar_t result = 0;
+  while (x < 10) {
+    result -= 1 / x;
+    x += 1;
+  }
+  if (x == 10) {
+    return result + PSI_10 + additional_summand;
+  }
+
+  // Compute asymptotic digamma
+  static const acc_scalar_t A[] = {
+     8.33333333333333333333E-2,
+    -2.10927960927960927961E-2,
+     7.57575757575757575758E-3,
+    -4.16666666666666666667E-3,
+     3.96825396825396825397E-3,
+    -8.33333333333333333333E-3,
+     8.33333333333333333333E-2,
+  };
+
+  acc_scalar_t y = 0;
+  if (x < 1.0e17) {
+    acc_scalar_t z = 1.0 / (x * x);
+    y = z * polevl<acc_scalar_t>(z, A, 6);
+  }
+  return static_cast<scalar_t>(result + std::log(x) - (0.5 / x) - y + additional_summand);
+}
+
+// Computes the reparameterized gradient -(d/dalpha cdf(x;alpha)) / pdf(x;alpha)
+// for random number x drawn from a standard Gamma distribution Gamma(alpha).
+template <typename scalar_t>
+deviceforcuda scalar_t standard_gamma_grad_one(scalar_t alpha, scalar_t x) {
+  // Use a Taylor series expansion for small x.
+  if (x < 0.8f) {
+    scalar_t numer = 1;
+    scalar_t denom = alpha;
+    auto series1 = numer / denom;
+    auto series2 = numer / (denom * denom);
+    for (int i = 1; i <= 5; ++i) {
+      numer *= -x / i;
+      denom += 1;
+      series1 += numer / denom;
+      series2 += numer / (denom * denom);
+    }
+    const auto pow_x_alpha = std::pow(x, alpha);
+    const auto gamma_pdf = std::pow(x, alpha - 1) * std::exp(-x);
+    const auto gamma_cdf = pow_x_alpha * series1;
+    const auto gamma_cdf_alpha = (std::log(x) - digamma_one(alpha)) * gamma_cdf
+        - pow_x_alpha * series2;
+    const auto result = -gamma_cdf_alpha / gamma_pdf;
+    return isnan(result) ? 0 : result;
+  }
+
+  // Use a Rice saddle point expansion for large alpha.
+  if (alpha > 8.0f) {
```

Comment:
This part is especially sensitive and probably needs to be computed with at least single precision.